src/test/csf/timers.h

```diff
@@ -30,6 +30,143 @@ namespace csf {
 
 // Timers are classes that schedule repeated events and are mostly independent
 // of simulation-specific details.
+/*
+    We should consider making DurationDistribution conform to
+    RandomNumberDistribution concept and moving to beast.
+
+    struct DurationDistribution
+    {
+        SimDuration
+        operator()();
+    };
+
+    Note T should be light-weight, trivially copyable and movable, etc.
+*/
+
+
+/*
+    ConstantDuration is a DurationDistribution
+*/
+struct ConstantDuration
+{
+    SimDuration value;
+
+    SimDuration
+    operator()()
+    {
+        return value;
+    }
+};
+
+
+/*
+    RandomDuration is a DurationDistribution which takes a Generator, passes it
+    through a RandomNumberDistribution to produce a random number, and then uses
+    it generate a Duration.
+*/
+template<class RandomNumberDistribution, class UniformRandomBitGenerator>
+struct RandomDuration
+{
+    RandomNumberDistribution dist;
+    UniformRandomBitGenerator & g;
+
+    //------------------------
+    // Convert generated durations to SimDuration
+    static SimDuration
+    asDuration(SimDuration d)
+    {
+        return d;
+    }
+
+    template <class T>
+    static
+    std::enable_if_t<std::is_arithmetic<T>::value, SimDuration>
+    asDuration(T t)
+    {
+        return SimDuration{static_cast<SimDuration::rep>(t)};
+    }
+
+    SimDuration
+    operator()()
+    {
+        return asDuration(dist(g));
+    }
+};
+
+template<class RandomNumberDistribution, class UniformRandomBitGenerator>
+RandomDuration<RandomNumberDistribution, UniformRandomBitGenerator>
+randomDuration(RandomNumberDistribution dist, UniformRandomBitGenerator& gen)
+{
+    return RandomDuration<RandomNumberDistribution, UniformRandomBitGenerator>{
+            dist, gen};
+};
+
+class DurationDistributionRef
+{
+    using tp = SimTime;
+    using dur = SimDuration;
+
+    struct IDurationDistribution
+    {
+        virtual ~IDurationDistribution() = default;
+
+        virtual SimDuration
+        operator()() = 0;
+    };
+
+    template <class T>
+    class Any final : public IDurationDistribution
+    {
+        T & t_;
+
+    public:
+        Any(T & t) : t_{t}
+        {
+        }
+
+        // Can't copy
+        Any(Any const & ) = delete;
+        Any& operator=(Any const & ) = delete;
+
+        Any(Any && ) = default;
+        Any& operator=(Any && ) = default;
+
+        virtual SimDuration
+        operator()()
+        {
+            return t_();
+        }
+    };
+
+    std::shared_ptr<IDurationDistribution> impl_;
+
+public:
+    template <class T>
+    explicit
+    DurationDistributionRef(T& t) : impl_{new Any<T>(t)}
+    {
+    }
+
+    template <class T>
+    explicit
+    DurationDistributionRef(T&& t) : impl_{new Any<T>(t)}
```

Comment:
I think this constructor might not do as you expect.   Since `Any<T>` is taking its argument by reference, it will end up referring to something that no longer exists.